Chapter 20 web api rest (#18)
LaunchCodeEducation/csharp-web-dev-curriculum: content/web-api-rest/studio/_index.md

draft: false
weight: 5
originalAuthor: John Woolbright # to be set by page creator
originalAuthorGitHub: jwoolbright23 # to be set by page creator
reviewer: Kimberly Horan # to be set by the page reviewer
reviewerGitHub: codinglikeagirl42 # to be set by the page reviewer
lastEditor: # update any time edits are made after review
lastEditorGitHub: # update any time edits are made after review
lastMod: 12/15/22 # UPDATE ANY TIME CHANGES ARE MADE
---

The UI of a browser is designed to make simple `GET` requests for URLs entered into its address bar. This design works great for browsing sites, but falls short when working with APIs. Anything beyond a `GET` request is difficult to send via a browser address bar alone. Think about what is needed to create a new `CodingEvent`. This type of request contains a body. Our MVC application included a view to allow us to test inputs. Our API, however, is **headless**. It does not contain the client-side form. In order to test how it handles requests then, we need a way to interact with the API server without the browser. In this studio, we work with Postman to explore how APIs can be consumed.

## Setup

### Install Postman

> If you haven't done so already, [install Postman](https://education.launchcode.org/csharp-web-dev-curriculum/appendices/install-postman/)

Reviewer: ```suggestion
If you haven't done so already, [install Postman]({{< relref "../../../appendices/install-postman/_index.md" >}}).
```

### Fork and Clone the API Source Code

With Postman installed, we're ready to fork and clone the [CodingEvents API](https://github.com/LaunchCodeEducation/coding-events-api/). Although it is implemented differently from CodingEvents MVC, you will find that most of the features from the MVC application have been supported through endpoints in the API.

{{% notice blue "Note" "rocket" %}}
Our focus in this course is on operations and as such we will not be going into the development of the API. However, feel free to explore the source code if you are curious about the similarities and differences between the .NET MVC and API implementations.
{{% /notice %}}

Let's begin by forking and cloning the repo onto our machine. In your Powershell or terminal window, move into a directory where you plan to save your local copy of the API codebase.

{{% notice blue "Note" "rocket" %}}
Including a SQLite database in this project means you don't need to have your MySQL server running to test the API. We won't get into what this looks like and instead just concentrate on testing the API endpoints.
{{% /notice %}}

> Let's change into the repo and switch to this branch:
> 
> ```bash
> # cd is an alias (like a nick-name) for the Set-Location cmdlet in PowerShell
> cd coding-events-api
> ```
> 
> Keep your terminal open! We have some more work to do.
> 

Reviewer: ```suggestion
```

### Start the API Server

We'll start the API server from the terminal using the `dotnet run` command. Navigate to the `CodingEventsAPI` project folder *within* your `coding-events-api` solution. This is the folder that contains `Controllers/` and so on, and is NOT the main project folder.

> ```bash
> # change to the CodingEventsAPI project directory
> cd CodingEventsAPI
> ```

Reviewer: ```suggestion
```bash
# navigate and change into the coding-events-api solution 
cd coding-events-api
# change to the CodingEventsAPI project directory
cd CodingEventsAPI
```
```

draft: false
weight: 5
originalAuthor: John Woolbright # to be set by page creator
originalAuthorGitHub: jwoolbright23 # to be set by page creator
reviewer: Kimberly Horan # to be set by the page reviewer
reviewerGitHub: codinglikeagirl42 # to be set by the page reviewer
lastEditor: # update any time edits are made after review
lastEditorGitHub: # update any time edits are made after review
lastMod: 12/15/22 # UPDATE ANY TIME CHANGES ARE MADE
---

The UI of a browser is designed to make simple `GET` requests for URLs entered into its address bar. This design works great for browsing sites, but falls short when working with APIs. Anything beyond a `GET` request is difficult to send via a browser address bar alone. Think about what is needed to create a new `CodingEvent`. This type of request contains a body. Our MVC application included a view to allow us to test inputs. Our API, however, is **headless**. It does not contain the client-side form. In order to test how it handles requests then, we need a way to interact with the API server without the browser. In this studio, we work with Postman to explore how APIs can be consumed.

## Setup

### Install Postman

> If you haven't done so already, [install Postman](postman-installation)

Reviewer: Add link to the new installing postman page.

Author: New page added and linked to.

### Fork and Clone the API Source Code

> With Postman installed, we're ready to fork and clone the [CodingEvents API](https://github.com/LaunchCodeEducation/coding-events-api/tree/1-sqlite). Although it is implemented differently from CodingEvents MVC, you will find that most of the features from the MVC application have been supported through endpoints in the API.

Reviewer: This links to the LaunchCode repo - the repo i used was this one ->

https://github.com/jwoolbright23/coding-events-api/tree/1-sqlite

Author: resolved link issues here

{{% notice blue "Note" "rocket" %}}
Our focus in this course is on operations and as such we will not be going into the development of the API. However, feel free to explore the source code if you are curious about the similarities and differences between the .NET MVC and API implementations.
{{% /notice %}}

> Let's begin by forking and cloning the repo onto our machine. In your Powershell or terminal window, move into a directory where you plan to save your local copy of the API codebase.
> 
> ```bash
> git clone https://github.com/<GitUsername>/coding-events-api
> ```

Reviewer: ```suggestion
```

Reviewer: Students should be able to do this without prompting, it is the last studio in the second unit.  Suggestion delete.

For this studio, we want to have the `1-sqlite` branch checked out. This branch has an API with a single (`CodingEvent`) 
resource and a built-in SQLite database. 

{{% notice blue "Note" "rocket" %}}
Including a SQLite database in this project means you don't need to have your MySQL server running to test the API. We won't get into what this looks like and instead just concentrate on testing the API endpoints.
{{% /notice %}}

> Let's change into the repo and switch to this branch:
> 
> ```bash
> # cd is an alias (like a nick-name) for the Set-Location cmdlet in PowerShell
> cd coding-events-api
> ```
> 
> ```bash
> # check out the 1-sqlite branch
> git checkout 1-sqlite
> ```
> 

Reviewer: ```suggestion
```

> Keep your terminal open! We have some more work to do.

Reviewer: Open your terminal and make sure you are in the correct repo and on the correct branch.

> Keep your terminal open! We have some more work to do.

Reviewer: ```suggestion
```

- **list CodingEvents**: `GET /api/events -> CodingEvent[]`
- **create a CodingEvent**: `POST /api/events (NewCodingEvent) -> 201, CodingEvent`

And two that require a sub-path variable, `/events/{codingEventId}`, to interact with a single entity:

- **find single CodingEvent**: `GET /api/events/{codingEventId} -> CodingEvent`
- **delete a CodingEvent**: `DELETE /api/events/{codingEventId} -> 201, CodingEvent`

And below this list are two *Schemas*, or shapes, of resources the API handles. 

**Swagger** is a popular tool API developers use to include fast documentation for their API codebase. The page we're looking at is generated by this tool and gives us a nice summary of the endpoints made available by the CodingEvents API we currently have running. But remember, we'll test the API in Postman.

## Making Requests to the Coding Events API

### List the Coding Events

To create our first request using Postman, select the *Create New* button located within the "Start with something new" section.

![Close up of the Postman create new button](pictures/create-new.png?classes=border)

Select the *Create New* button to create a new request

{{% notice blue "Note" "rocket" %}}

> The above screenshot and the screenshots on this page are of the Linux version of Postman. The *Create New* button on your launchpad should also be located under the "Start with something new" section of the home page.

Reviewer: ```suggestion
The above screenshot is from the Home page on Postman. The *Create New* button on your launchpad should also be located under the "Start with something new" section.  If you are on the Workshop page choose the *New* button on the top right of the *My Workspace* pane located on the left side.
 ```
